src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4822,6 +4573,42 @@ private static void AddUniqueVariable(HashSet<string> hashedResults, List<Comple
             }
         }
 
+        private static string GenerateVariableCompletionText(string value)
+        {
+            return GenerateVariableCompletionText(value, true);
+        }
+
+        private static string GenerateVariableCompletionText(string value, bool hasProvider)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
+
+            bool startsWithQM = value[0] == '?' && value.Length > 1;
+            bool lastCharWasColon = false, hasColon = false;
+            bool braceVariable = !value[0].IsVariableStart() || value.Length != 1 && (value[0] == '$' || value[0] == '^' || (hasProvider && startsWithQM));
+            if (!braceVariable)
+            {
+                lastCharWasColon = !hasProvider && value[0] == ':';
+                hasColon = lastCharWasColon;
+                foreach (char c in value.Substring(1))
+                {
+                    if ((c == ':' && lastCharWasColon) || !(c.IsIdentifierFollow() || c == '?' || c == ':'))
+                    {
+                        braceVariable = true;
+                        break;
+                    }
+                    lastCharWasColon = c == ':';
+                    hasColon |= lastCharWasColon;
+                }
+            }
+
+            // note add a leading `:` if no provider is part of the completion, but completion contains `:`
+            // or specifically for an unbraced variable that starts with a `?` (but is more than just a `?`)
+            return braceVariable ?
+                "{" + CodeGeneration.EscapeVariableName((hasProvider || !hasColon ? string.Empty : ":") + value) + "}" :
+                (hasProvider || !(startsWithQM || hasColon) ? string.Empty : ":") + value;
```

Comment:
Can we simplify the return? It is difficult to read.

Reply:
That's why there is a set of comment lines right before it.  The expression cannot get 'simpler'.  It can be expanded in to the 4 different return expressions it represents.  while each of the expanded return expression will appear simpler, there will be 4 expressions, and three tests.  

When evaluated closely, there are two main returns, and I think the first ternary makes that easy to see; each of the main returns is on its own line, and then in each one is a unique decisions to insert a ':' in front of the variable name being offered for completion so that the completion will work as expected.

When you add the 'prefix +' to the expression it makes the single return statement that much more efficient:
```C#
            return prefix + (braceVariable ?
                "{" + CodeGeneration.EscapeVariableName((hasProvider || !hasColon ? string.Empty : ":") + value) + "}" :
                (hasProvider || !(startsWithQM || hasColon) ? string.Empty : ":") + value);
```

Reply:
I meant:
```c#
if (braceVariable)
{
    return …
}
else
{
    return …
}
```